### PR TITLE
perf(core): scoped incremental cache invalidation (#1904)

### PR DESCRIPTION
**Pull request by joshuaswarren**
Part of #1904.

Replaces the single all-layers `invalidateAllForDir()` (fired on every memory write) with SCOPE-AWARE invalidation, so a fact `create` no longer nukes the QMD recall/search caches, the entity/Knowledge-Index cache, or the archive cache — only the hot + derived views that actually change. Mutate paths (update/supersede/archive/invalidate/frontmatter/cold/bulk) and entity writes still evict the QMD caches (lifecycle-filter correctness preserved). Plus an incremental GraphIndex edge cache (append in place + file-size revalidation).

- **Scope taxonomy** (`memory-cache.ts`): `MemoryMutationScope` (memory-create | memory-mutate | entity-write | secure-key-change) + `evictOn` per `ALL_CACHE_LAYERS` entry per the issue matrix; `invalidateForScope` added, `invalidateAllForDir` kept as the union.
- **Routing** (`storage.ts`): `writeMemory`/`writeChunk` creates → `patchHotMemoriesCache(..., "memory-create")` (in-place hot patch preserved from #1902; QMD/entity/archive stay warm); `invalidateAllMemoriesCache` → `invalidateForScope("memory-mutate")` (guarded). `bumpMemoryStatusVersion`/`setSecureStoreKey`/`invalidateColdMemoriesCache` stay full clear (correctness backstops).
- **Incremental edge cache** (`graph.ts`): record per-file byte sizes; `onMemoryWritten` pushes appended edges in place when `edgeFilesUnchangedExceptOurAppends` confirms sole-writer coherence (exact `${JSON.stringify(edge)}\n` byte delta), else nulls (full reload). 5-min TTL backstop kept.
- **Config**: `scopedCacheInvalidationEnabled` + `graphEdgeCacheIncrementalEnabled` (both default `true`; `false` restores byte-for-byte legacy). Declared in both openclaw manifests (byte-identical) + root manifest.

Guardrails honored: recall pipeline order/results unchanged; superseded/deleted/archived/status-changed never resurface; cross-process coherence preserved (QMD cleared on visibility changes; entity cache version-keyed; edge cache size-revalidated); zero-limit/`enabled=false` restore legacy; fail-open; single-chokepoint invariant intact. Rewrote `tests/cache-invalidation-coherence.test.ts` to the scope matrix + added create/entity/graph/config tests.

---

> [!NOTE]
> **Medium Risk**
> Changes cache invalidation and graph/QMD coherence paths used on every write and recall; correctness is heavily tested but stale-cache edge cases under concurrent writers remain TTL-bounded.
> 
> **Overview**
> **Scope-aware invalidation (#1904)** replaces clearing every cache layer on each write with mutation scopes (`memory-create`, `memory-mutate`, `entity-write`, `secure-key-change`) and per-layer `evictOn` rules. Fact **creates** (`writeMemory` / `writeChunk`) still patch hot memories in place but only refresh derived episode/rule views—**QMD search/recall**, **entity**, and **archive** caches stay warm. **Mutates** still evict QMD caches so stale recall bundles cannot resurface (#1535). `scopedCacheInvalidationEnabled` (default `true`) restores full-clear-on-write when set `false`.
> 
> **QMD coherence:** successful `update` / `embed` (and dimension-mismatch recovery) now call `clearQmdResultCaches()` centrally in `QmdClient`, so caches can stay warm during index lag but invalidate as soon as indexing actually advances.
> 
> **Incremental graph edge cache:** `onMemoryWritten` appends edges into the warm cache when file identity (size, mtime, inode) proves sole-writer coherence; peer writes or atomic rewrites force a reload. `graphEdgeCacheIncrementalEnabled` (default `true`) restores null-on-every-write when `false`.
> 
> Tests expand `cache-invalidation-coherence` to the scope matrix; graph tests cover incremental vs legacy behavior. Plugin manifests and config parsing expose both flags.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit c2425d9e999e944b63c321afe668227a3964e200. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/remnic-core/src/storage.ts [hunk 1]
```diff
@@ -4684,9 +4721,19 @@ export class StorageManager {
...
+    // Evict the non-hot layers; the hot layer is patched (or intentionally left
+    // to be version-rejected) above, not dropped here. Scope-aware (issue #1904):
+    // a `memory-create` evicts ONLY the derived episode/rule views, keeping the
+    // global QMD result caches and the version-keyed entity cache warm (a create
+    // adds a doc that is not yet in the QMD index and changes no existing doc's
+    // recall visibility). A `memory-mutate` (or scoped-off rollback) still evicts
+    // the derived + global QMD views so a superseded/edited memory cannot resurface
+    // from a stale recall bundle (issue #1535 correctness contract).
+    if (scope === "memory-create" && (StorageManager.scopedCacheInvalidationByDir.get(this.baseDir) ?? true)) {
+      invalidateForScopeExceptHot(this.baseDir, "memory-create");
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/storage.ts:4733]: Fixed in c7cdfbc0 — runQmdMaintenance now clears the QMD result caches (qmd-search + qmd-recall) after a successful update+embed. A newly-persisted fact becomes searchable at that moment, so any cached pre-index bundle is invalidated then — the create path keeps the QMD caches warm during index lag (the perf win) but there is no extended stale window past indexing. Bounded by the existing fresh/stale TTLs as a backstop.

File: packages/remnic-core/src/graph.ts [hunk 2]
```diff
@@ -444,61 +486,128 @@ export class GraphIndex {
...
+    if (this.cfg.entityGraphEnabled) enabled.push("entity");
+    if (this.cfg.timeGraphEnabled) enabled.push("time");
+    if (this.cfg.causalGraphEnabled) enabled.push("causal");
+    const observed: Partial<Record<GraphType, number>> = {};
+    for (const type of enabled) {
+      let size: number;
+      try {
+        size = (await stat(graphFilePath(this.memoryDir, type))).size;
+      } catch {
+        return false;
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/graph.ts]: Fixed in c7cdfbc0 — edgeFilesUnchangedExceptOurAppends now treats an ENOENT graph file as its recorded 0-byte baseline (the normal partially-populated state, e.g. entity edges written before any causal edge) instead of returning false. Only a genuine stat failure forces a full reload, so the incremental edge cache stays warm for the common multi-graph state.

File: packages/remnic-core/src/graph.ts [hunk 3]
```diff
@@ -401,10 +421,32 @@ export class GraphIndex {
       timeGraph: this.cfg.timeGraphEnabled,
       causalGraph: this.cfg.causalGraphEnabled,
     });
-    this.edgeCache = { allEdges, loadedAt: Date.now() };
+    this.edgeCache = { allEdges, loadedAt: Date.now(), sizes: await this.readEnabledGraphFileSizes() };
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/graph.ts]: Fixed in 1e2b6fc4 — the edge cache now records a full {size, mtimeMs, ino} identity per file, not just size, and loadEdgesCached() captures that identity BEFORE and AFTER readAllEdges() with a bounded retry, so allEdges and the baseline reflect the same on-disk state. A peer append racing construction now either bumps the between-snapshot size (retry) or fails the validation size delta (reload); it can no longer leave the array missing a peer edge while the baseline counts its bytes.

File: packages/remnic-core/src/graph.ts [hunk 4]
```diff
@@ -444,61 +486,136 @@ export class GraphIndex {
...
+        // (e.g. entity edges written before any causal edge): treat it as its
+        // recorded 0-byte baseline so the incremental cache stays warm. Only a
+        // genuine stat failure forces a full reload (#1904, Codex).
+        if ((err as NodeJS.ErrnoException)?.code === "ENOENT") {
+          size = 0;
+        } else {
+          return false;
+        }
+      }
+      if (size !== (this.edgeCache.sizes[type] ?? 0) + (expectedDelta[type] ?? 0)) return false;
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/graph.ts]: Fixed in 1e2b6fc4 — the identity includes the inode. A local append (appendFile) keeps the inode; the decay writer's atomic temp+rename changes it, so an equal-length rewrite (e.g. confidence 0.9->0.8) now diverges on inode and forces a reload. Untouched files must also match size AND mtime, so an in-place equal-length rewrite (mtime bump) is caught as well. Regression test added: onMemoryWritten reloads when an equal-length atomic rewrite changed the inode.

File: packages/remnic-core/src/orchestration/maintenance.ts [hunk 5]
```diff
@@ -523,6 +524,11 @@ export class MaintenanceScheduler {
           this.lastQmdEmbedAtMs = now;
         }
       }
+      // A successful update+embed means newly-persisted facts are now searchable;
+      // any cached pre-index QMD recall/search bundle is now stale. Clear ONLY the
+      // QMD result caches (dir-scoped layers untouched) so the create path can keep
+      // them warm during index lag without an extended stale window (#1904, Codex).
+      clearQmdResultCaches();
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/orchestration/maintenance.ts]: Fixed in 3765f4fe — runQmdMaintenance now tracks a didUpdate flag. The namespace path sets it from runNamespaceMaintenanceBatchPlan's summary.ran > 0 (a fully-skipped throttle/lock_held pass has ran=0), the non-namespace path sets it after getQmd().update(). clearQmdResultCaches() runs only when didUpdate is true, so a skipped pass keeps the warm caches and the two paths no longer diverge.

File: packages/remnic-core/src/storage.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/storage.ts:4733]: Fixed in 3765f4fe — the wearable sync reindex path (getWearablesService reindexSearch, called via wearables/pipeline.ts afterWrites) now calls clearQmdResultCaches() right after its direct qmd.update(), so newly-indexed transcript/fact content is not hidden by a cache entry warmed before the sync.

File: packages/remnic-core/src/graph.ts [hunk 6]
```diff
@@ -396,15 +433,71 @@ export class GraphIndex {
...
+        timeGraph: this.cfg.timeGraphEnabled,
+        causalGraph: this.cfg.causalGraphEnabled,
+      });
+      meta = await this.readEnabledGraphFileMeta();
+      if (this.graphFileMetaEqual(before, meta)) break;
+      // A write landed during the read; the just-captured `meta` is the new
+      // baseline for the next attempt's read. After the final attempt we accept
+      // the last read paired with its own post-read `meta` (bounded by the TTL).
+    }
+    this.edgeCache = { allEdges, loadedAt: Date.now(), meta };
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/graph.ts]: Fixed in 3765f4fe — loadEdgesCached tracks a stable flag; if no before/after snapshot matched across the 3 bounded retries, it leaves edgeCache null (serving this read's edges as best-effort) instead of installing the final inconsistent allEdges/meta pair, so a continuously-writing peer cannot poison the cache baseline and hide its edge for the TTL.

File: packages/remnic-core/src/graph.ts [hunk 7]
```diff
@@ -444,61 +549,164 @@ export class GraphIndex {
...
+        this.edgeCache = null;
+      } else {
+        // Trust an in-place push ONLY when WE are the sole writer since load:
+        // every enabled graph file's on-disk size must equal our baseline plus
+        // exactly the bytes we appended to it. A divergence means a peer process
+        // (backup daemon / CLI / decay job) also wrote — force a full reload for
+        // cross-process coherence. Any stat error fails open to a full reload.
+        try {
+          if (await this.edgeFilesUnchangedExceptOurAppends(appended)) {
+            if (appended.length > 0) this.edgeCache.allEdges.push(...appended);
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/graph.ts]: Fixed in 8e4f3c49 — after the in-place push, onMemoryWritten now re-stats every committed graph file (edgeCacheIdentityUnchangedSinceCommit) and nulls the cache if any identity moved, so a peer append in the validation→push window cannot leave a stale cache for the TTL. The re-verify runs only when this call actually pushed edges (no new window otherwise; the validation already confirmed no peer wrote). A residual infinitesimal window during the re-stat itself remains and is TTL-bounded.

File: packages/remnic-core/src/orchestration/maintenance.ts [hunk 8]
```diff
@@ -522,7 +528,14 @@ export class MaintenanceScheduler {
...
           this.lastQmdEmbedAtMs = now;
         }
+        didUpdate = true;
       }
+      // A successful update+embed means newly-persisted facts are now searchable;
+      // any cached pre-index QMD recall/search bundle is now stale. Clear ONLY the
+      // QMD result caches (dir-scoped layers untouched) — but only when an index
+      // actually changed, so a fully-skipped (throttled) pass keeps them warm
+      // (#1904, Codex).
+      if (didUpdate) clearQmdResultCaches();
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/orchestration/maintenance.ts]: Fixed in cd62ce0d — didUpdate is now derived from the backend's lastUpdateRanAtMs getter (exposed as an optional readonly member on SearchBackend; QmdClient already tracks it). The non-namespace path snapshots it before update() and sets didUpdate iff it advanced, so a no-op pass (unavailable / min-interval / failure backoff — all of which return silently without advancing the timestamp) keeps the warm QMD caches. An absent getter falls back to 'assume refreshed' so other backends are unaffected.

[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/orchestration/maintenance.ts]: Fixed in cd62ce0d — didUpdate is now assigned immediately after update() succeeds (from the lastUpdateRanAtMs advancement), before the optional embed(), so an embed() rejection no longer skips the invalidation: update() had already indexed the fact, didUpdate is true, and clearQmdResultCaches() runs. The embed() is additionally gated on didUpdate, so a no-op update also skips a wasteful embed.

File: packages/remnic-core/src/orchestration/maintenance.ts [hunk 9]
```diff
@@ -512,17 +514,38 @@ export class MaintenanceScheduler {
             skipReasonForError: qmdMaintenanceSkipReasonForError,
           },
         );
+        // Only a real per-namespace update makes new facts searchable; if every
+        // namespace was skipped (throttle / lock_held) no index changed, so there
+        // is nothing to invalidate and we keep the warm QMD caches (#1904, Codex).
+        didUpdate = summary.ran > 0;
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/orchestration/maintenance.ts]: Fixed centrally in 67d6e655 — QmdClient.runUpdateForCollection now clears the global QMD result caches the moment a refresh succeeds, BEFORE any embed runs. Since the namespace router delegates every per-namespace update to backend.update(), this covers the namespace path too, and an embedNamespaces() rejection can no longer skip the invalidation: updateNamespacesDetailed already indexed (clearing the caches) before embed is even attempted. The didUpdate/summary.ran tracking that was subject to this race is removed.

File: packages/remnic-core/src/storage.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/storage.ts:4733]: Fixed centrally in 67d6e655 — the OpenClaw provider sync route (src/index.ts) calls orchestrator.qmd.update() directly, which is QmdClient.update(); that now clears the global QMD result caches on success inside runUpdateForCollection. So this and every other direct-update path (wearable reindex, CLI, namespace router) invalidate centrally without each having to remember — no more whack-a-mole.
